### PR TITLE
Fix computing of array element type for `∇eachslice`

### LINKED ISSUE #807: `∇eachslice` causes `UndefRefError` when input array contains references
https://github.com/JuliaDiff/ChainRules.jl/blob/30f9b121e69fe20f3a176e92e3ceb91f53606f8c/src/rulesets/Base/indexing.jl#L264-L282

1. `similar` makes `dx` an uninitialized array.
2. When `dys[i]` is zero, presumably the intent of `_zero_fill!(slice)` is to populate the array with zeros.
3. `_zero_fill!` on an uninitialized array calls `map!(zero, dx, dx)`
4. `map!` attempts to map `zero` onto the elements of `dx`, which are `#undef` - error

So this branch never works when the input array has references rather than inline values.

### PR DESCRIPTION
**Pull request by BioTurboNick**
Fixes #807

Added test as well. Two lines really belong in ChainRulesCore, not sure what the process for that is.

### CODE REVIEW

**Comment by BioTurboNick** [comment]
Other than the formatter, all the failures seem to be due to the fact that the changes mean Julia currently isn't fully inferring the output anymore. I'll see if there's anything to do about that.

**Comment by BioTurboNick** [comment]
Switching to `Base.promote_eltype` addresses the inference issue, but that's not a public API in base.

**Comment by BioTurboNick** [comment]
Huh. I'm not sure why the inference test is failing on 1.6 only. It's inferring `Union{NoTangent, Vector{Float64})` instead of just Vector{Float64}'.

File: src/rulesets/Base/indexing.jl [hunk 1]
```diff
@@ -282,8 +282,12 @@ function ∇eachslice(dys_raw, x::AbstractArray, vd::Val{dim}) where {dim}
...
 ∇eachslice(dys::AbstractZero, x::AbstractArray, vd::Val{dim}) where {dim} = dys
 
-_zero_fill!(dx::AbstractArray{<:Number}) = fill!(dx, zero(eltype(dx)))
-_zero_fill!(dx::AbstractArray) = map!(zero, dx, dx)
+_zero_fill!(dx::AbstractArray) = fill!(dx, zero(eltype(dx)))
+
+# Belong in ChainRulesCore
+Base.promote_rule(T::Type{<:Number}, S::Type{<:AbstractZero}) = T
+Base.promote_rule(T::Type{<:AbstractZero}, S::Type{<:Number}) = S
+Base.eltype(::Type{NoTangent}) = NoTangent
```
Comment (simeonschaub) [inline src/rulesets/Base/indexing.jl:290]: Can you make a PR to ChainRulesCore?

**Review by simeonschaub — CHANGES_REQUESTED** [review]
I believe the original intent with the design of `similar` was to support cases like `Vector{Vector{Float64}}` where `zero(Vector{Float64})`. That of course doesn't work though since `similar` will just produce `#undef`s here making this fail anyways.

Looks good to me minus the methods that need to be moved to ChainRulesCore. Could you make a PR there (ideally with some unit tests), so we can tag a new version there and then change the compat here.

File: src/rulesets/Base/indexing.jl (hunk 1 above)
Comment (BioTurboNick) [reply src/rulesets/Base/indexing.jl:290]: Done!

**Comment by simeonschaub** [comment]
Just needs to drop the changes already in https://github.com/JuliaDiff/ChainRulesCore.jl/pull/682 and bump the compat for ChainRulesCore

**Comment by BioTurboNick** [comment]
done and done

**Comment by simeonschaub** [comment]
Hmm, looks like the format suggestions are not working properly (Maybe because it's on a fork?) If you look at the CI logs, you can see that it complains about the formatting of line 257 in the tests. Once that's fixed this looks good to go!

**Comment by devmotion** [comment]
> (Maybe because it's on a fork?) 

Yes, due to restricted permissions for forks the suggestions can only be shown in the logs but not as actual suggestions in the PR.

**Review by simeonschaub — APPROVED** [review]

**Pull request merged**